shuffle-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/ucx/UCXShuffleTransport.scala

```diff
@@ -160,84 +170,43 @@ class UCXShuffleTransport(shuffleServerId: BlockManagerId, rapidsConf: RapidsCon
     Seq(deviceSendBuffMgr, deviceReceiveBuffMgr, hostSendBuffMgr).foreach(_.close())
   }
 
-  def freeReceiveBounceBuffers(bounceBuffers: Seq[MemoryBuffer]): Unit = {
-    bounceBuffers.foreach(bb => {
-      deviceReceiveBuffMgr.freeBuffer(bb)
-    })
-
-    // let the throttle know some bounce buffers were freed, so we give it a chance to claim them
-    inflightMonitor.synchronized {
-      inflightMonitor.notify()
-    }
-  }
-
-  def freeSendBounceBuffers(bounceBuffers: Seq[MemoryBuffer]): Unit = {
-    bounceBuffers.foreach {
-      case hb: HostMemoryBuffer => hostSendBuffMgr.freeBuffer(hb)
-      case db: DeviceMemoryBuffer => deviceSendBuffMgr.freeBuffer(db)
-    }
-  }
-
   private def getNumBounceBuffers(remaining: Long, totalRequired: Int): Int = {
     val numBuffers = (remaining + bounceBufferSize - 1) / bounceBufferSize
     Math.min(numBuffers, totalRequired).toInt
   }
 
-  override def getSendBounceBuffers(
-      deviceMemory: Boolean,
-      remaining: Long,
-      totalRequired: Int): Seq[MemoryBuffer] = {
-
-    val numBuffs = getNumBounceBuffers(remaining, totalRequired)
-    if (!deviceMemory) {
-      acquireBounceBuffers(hostSendBuffMgr, numBuffs)
-    } else {
-      acquireBounceBuffers(deviceSendBuffMgr, numBuffs)
-    }
-  }
-
   override def tryGetSendBounceBuffers(
-      deviceMemory: Boolean,
       remaining: Long,
-      totalRequired: Int): Seq[MemoryBuffer] = {
-
+      totalRequired: Int): Seq[SendBounceBuffers] = {
     val numBuffs = getNumBounceBuffers(remaining, totalRequired)
-    if (!deviceMemory) {
-      tryAcquireBounceBuffers(hostSendBuffMgr, numBuffs)
+    val deviceBuffer = tryAcquireBounceBuffers(deviceSendBuffMgr, numBuffs)
+    if (deviceBuffer.nonEmpty) {
+      val hostBuffer = tryAcquireBounceBuffers(hostSendBuffMgr, numBuffs)
+      if (hostBuffer.nonEmpty) {
+        deviceBuffer.zip(hostBuffer).map { case (d, h) =>
```

Comment:
If I understand the code correctly, `tryAcquireBounceBuffers` can return fewer buffers than `numBuffs`.  If that's the case, how to we know we acquired the same number of host buffers as device buffers?  If one had excess, how do we release the excess?

Reply:
I think this can be cleaned up, but right now the logic with 1 bounce buffer (hard coded) we are acquiring passes through this:

```
val possibleNumBuffers = Math.min(bounceBuffMgr.numBuffers, numBuffs)
```

So we end up with either 1 or 0. That's why I am not seeing this be an issue in this case, we'll get a device buffer, and we are likely to get a host buffer, unless the user configured the host buffer pool to have less buffers.

In a follow up PR, I am planning on cleaning this up, I can do it here, but I am not sure if you'd rather.

Reply:
Let me punt on this one for this PR. The caller to this function is hardcoding a 1 in the request for bounce buffers, so it will not be the case, even if the user misconfigures this, that we get extra host buffers. I am changing this actively so I do expect this to be cleared up.